### ISSUE TITLE
Using multiple GPUs for training on a single machine?

### ISSUE THREAD

**Issue by xxww** [tensorflow/lingvo#202]
I am trying to train a simple 34-layer Resnet model with the Imagenet dataset on a machine with multiple GPU cards (all v100s). I am using Lingvo-0.6.2 synced from github with TF 2.1.0 with Ubuntu 18.04.

I initially tried this setting: "--mode=sync --worker_gpus=2 --worker_split_size=2"
From nvidia-smi, I can see that both GPUs are used by Lingvo. But I am getting exactly the same speed as what I got from a single GPU. So basically it worked, but came with no performance gain. I also checked the other thread, and I saw CPUs are pretty idle, so definitely it is not bottlenecked at the CPU tasks such as reading examples.

Then I tried this setting: "--mode=async --worker_gpus=2 --worker_split_size=2"
Basically I switched to async training. And I got following error at the end. And I checked the source codes, it seems like for async, I will have to set up each job myself?

This is my first time using Lingvo and I hope you could offer some help here.

Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1855, in <module>
    tf.app.run(main)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1847, in main
    RunnerManager(FLAGS.model).Start()
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1843, in Start
    self.StartRunners(self.CreateRunners(FLAGS.job.split(','), FLAGS.logdir))
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1590, in CreateRunners
    trial)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1545, in _CreateRunner
    cfg = self.GetParamsForDataset('trainer_client', 'Train')
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1399, in GetParamsForDataset
    with cluster_factory.Cluster(cluster.params):
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/cluster.py", line 204, in __init__
    assert False, (p.mode, p.job)
AssertionError: ('async', 'trainer_client')

**Comment by jonathanasdf** [comment]
tl;dr you want worker_split_size=1

A split is one minibatch of inputs. With worker_split_size=2 you're saying that each minibatch should be given two gpus, but in this case the model you're using needs to explicitly set the devices to take advantage of the two gpus (example bidi-rnn where the forward and backward rnns are on different devices https://github.com/tensorflow/lingvo/blob/ac6adce5ba868e45f115781bb74001db26ce0195/lingvo/core/rnn_layers.py#L476)

The number of minibatches being processed concurrently will be worker_replicas * worker_gpus / worker_split_size.

**Comment by xxww** [comment]
Thanks for the response.

I tried the worker_split_size=1 trick.

So be specific:
My whole command line is as such:

python -m lingvo.trainer \                                                      
--logdir=$LOG_DIR \                                                           
--model=imagenet.Imagenet \                                                   
--resnet_depth=34 \                                                           
--run_locally=gpu \                                                           
--tfrecord_pattern=$TFRECORD \             
--mode=async \                                                                                                   
--worker_gpus=2 \                                                             
--worker_split_size=1 \                                                       

And I also tried both mode=sync and mode=async with worker_split_size=1.

When "mode=sync, worker_gpus=2, worker_split_size=1", I ran into this issue:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1619, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: slice index 0 of dimension 0 out of bounds. for 'strided_slice' (op: 'StridedSlice') with input shapes: [0], [1], [1], [1] and with computed input tensors: input[1] = <0>, input[2] = <1>, input[3] = <1>.

When "mode=async, worker_gpus=2, worker_split_size=1", I ran into the same issue as before:

Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1855, in <module>
    tf.app.run(main)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1847, in main
    RunnerManager(FLAGS.model).Start()
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1843, in Start
    self.StartRunners(self.CreateRunners(FLAGS.job.split(','), FLAGS.logdir))
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1590, in CreateRunners
    trial)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1545, in _CreateRunner
    cfg = self.GetParamsForDataset('trainer_client', 'Train')
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1399, in GetParamsForDataset
    with cluster_factory.Cluster(cluster.params):
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/cluster.py", line 204, in __init__
    assert False, (p.mode, p.job)
AssertionError: ('async', 'trainer_client')
ERROR:tensorflow:==================================

**Comment by jonathanasdf** [comment]
For the sync case, can you paste more of the error (eg. where the slice operation is defined)

For the async case, I think there is a bug here

https://github.com/tensorflow/lingvo/blob/1aba0c93ae9592af88e43460aad9d19fa5b87e5a/lingvo/trainer.py#L1677

Should be

```
else if FLAGS.mode == 'async':
  FLAGS.job = 'controller,trainer'
else:
  FLAGS.job = 'controller,trainer_client'
```

You can try to fix this locally.

**Comment by xxww** [comment]
I will patch your change in async asap. Meantime, here comes the full error log for "mode=sync":

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1619, in _create_c_op
    c_op = c_api.TF_FinishOperation(op_desc)
tensorflow.python.framework.errors_impl.InvalidArgumentError: slice index 0 of dimension 0 out of bounds. for 'strided_slice' (op: 'StridedSlice') with input shapes: [0], [1], [1], [1] and with computed input tensors: input[1] = <0>, input[2] = <1>, input[3] = <1>.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1855, in <module>
    tf.app.run(main)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1847, in main
    RunnerManager(FLAGS.model).Start()
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1843, in Start
    self.StartRunners(self.CreateRunners(FLAGS.job.split(','), FLAGS.logdir))
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1590, in CreateRunners
    trial)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1549, in _CreateRunner
    return self.Trainer(cfg, *common_args)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 391, in __init__
    self._model.ConstructFPropBPropGraph()
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/base_model.py", line 1169, in ConstructFPropBPropGraph
    self._task.FPropDefaultTheta()
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/base_model.py", line 522, in FPropDefaultTheta
    input_batch = self.GetInputBatch()
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/base_model.py", line 517, in GetInputBatch
    self.cluster.num_splits_per_client)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/cluster.py", line 369, in SplitInputBatch
    return super(_UseInputDevice, self).SplitInputBatch(num_splits)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/base_input_generator.py", line 308, in SplitInputBatch
    field_split = ig_helper.SplitTensors(batch.Flatten(), num_splits)
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/input_generator_helper.py", line 75, in SplitTensors
    batch_dims = [tf.shape(x)[0] for x in xs]
  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/input_generator_helper.py", line 75, in <listcomp>
    batch_dims = [tf.shape(x)[0] for x in xs]
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/array_ops.py", line 898, in _slice_helper
    name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/array_ops.py", line 1064, in strided_slice
    shrink_axis_mask=shrink_axis_mask)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/gen_array_ops.py", line 9535, in strided_slice
    shrink_axis_mask=shrink_axis_mask, name=name)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/op_def_library.py", line 742, in _apply_op_helper
    attrs=attr_protos, op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 3322, in _create_op_internal
    op_def=op_def)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1786, in __init__
    control_input_ops)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py", line 1622, in _create_c_op
    raise ValueError(str(e))
ValueError: slice index 0 of dimension 0 out of bounds. for 'strided_slice' (op: 'StridedSlice') with input shapes: [0], [1], [1], [1] and with computed input tensors: input[1] = <0>, input[2] = <1>, input[3] = <1>.
ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Tensor'>):
<tf.Tensor 'report_uninitialized_variables/boolean_mask/GatherV2:0' shape=(?,) dtype=string>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1540, in _CreateRunner
    return self.Controller(cfg, *common_args)  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 298, in __init__
    'train.pbtxt')  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 302, in _CreateCheckpointer
    return checkpointer.Checkpointer(train_dir, model)  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/checkpointer.py", line 70, in __init__
    tf.global_variables())  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/util/tf_should_use.py", line 237, in wrapped
    error_in_function=error_in_function)
==================================
E0408 01:55:46.862659 139879080245056 tf_should_use.py:92] ==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Tensor'>):
<tf.Tensor 'report_uninitialized_variables/boolean_mask/GatherV2:0' shape=(?,) dtype=string>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 1540, in _CreateRunner
    return self.Controller(cfg, *common_args)  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 298, in __init__
    'train.pbtxt')  File "/usr/local/lib/python3.6/dist-packages/lingvo/trainer.py", line 302, in _CreateCheckpointer
    return checkpointer.Checkpointer(train_dir, model)  File "/usr/local/lib/python3.6/dist-packages/lingvo/core/checkpointer.py", line 70, in __init__
    tf.global_variables())  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/util/tf_should_use.py", line 237, in wrapped
    error_in_function=error_in_function)
==================================

**Comment by xxww** [comment]
I can confirm that after packing the the mode=async code path, it will end up with the same error as mode=sync above.

**Comment by jonathanasdf** [comment]
It looks like it's trying to split the input batch onto the two devices and failing.

Can you make sure that everything in the input batch being returned from your input generator has a leading batch dimension?

One way to check is to add here
https://github.com/tensorflow/lingvo/blob/1aba0c93ae9592af88e43460aad9d19fa5b87e5a/lingvo/core/base_input_generator.py#L328

```python
for k, v in batch.FlattenItems():
  print(k, v.shape)
```

**Comment by xxww** [comment]
Thank you so much. Upon printing these tensors, I figure out I have the following offender:

  def InputBatch(self):
    batch = py_utils.NestedMap()
    batch.bucket_keys = 1  # commenting this out will fix it
    batch.rgb = self._rgb
    batch.label = self._label
    return batch
